diff --git a/build/npm/preinstall.ts b/build/npm/preinstall.ts
index cd5b63a1..fd7e684c 100644
--- a/build/npm/preinstall.ts
+++ b/build/npm/preinstall.ts
@@ -135,22 +135,22 @@ function installHeaders() {
 
 	if (local !== undefined) {
 		// Both disturl and target come from a file checked into our repository
-		child_process.execFileSync(node_gyp, ['install', '--dist-url', local.disturl, local.target], { shell: true });
+		child_process.execFileSync(node_gyp, ['install', '--dist-url', local.disturl, local.target], { shell: process.platform === 'win32' });
 	}
 
 	if (remote !== undefined) {
 		// Both disturl and target come from a file checked into our repository
-		child_process.execFileSync(node_gyp, ['install', '--dist-url', remote.disturl, remote.target], { shell: true });
+		child_process.execFileSync(node_gyp, ['install', '--dist-url', remote.disturl, remote.target], { shell: process.platform === 'win32' });
 	}
 
 	// On Linux, apply a patch to the downloaded headers
 	// Remove dependency on std::source_location to avoid bumping the required GCC version to 11+
 	// Refs https://chromium-review.googlesource.com/c/v8/v8/+/6879784
-	if (process.platform === 'linux') {
+	if (process.platform === 'linux' && local) {
 		const homedir = os.homedir();
 		const cachePath = process.env.XDG_CACHE_HOME || path.join(homedir, '.cache');
 		const nodeGypCache = path.join(cachePath, 'node-gyp');
-		const localHeaderPath = path.join(nodeGypCache, local!.target, 'include', 'node');
+		const localHeaderPath = path.join(nodeGypCache, local.target, 'include', 'node');
 		if (fs.existsSync(localHeaderPath)) {
 			console.log('Applying v8-source-location.patch to', localHeaderPath);
 			try {
